data-pipeline-ffi/src/error.rs

```diff
@@ -0,0 +1,161 @@
+            Self::InvalidArgument => write!(f, "Invalid argument provided"),
+            Self::InvalidData => write!(f, "Invalid data payload"),
+            Self::InvalidInput => write!(f, "Invalid input"),
+            Self::InvalidUrl => write!(f, "Invalid URL"),
+            Self::IoError => write!(f, "Input/Output error"),
+            Self::NetworkUnknown => write!(f, "Unknown network error"),
+            Self::Serde => write!(f, "Serialization/Deserialization error"),
+            Self::TimedOut => write!(f, "Operation timed out"),
+        }
+    }
+}
+
+/// Stucture that contains error information that `TraceExporter` API can return.
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+pub struct ExporterError {
+    pub code: ExporterErrorCode,
+    pub msg: *mut c_char,
+}
+
+impl ExporterError {
+    pub fn new(code: ExporterErrorCode, msg: &str) -> Self {
+        Self {
+            code,
+            msg: CString::new(msg).unwrap_or_default().into_raw(),
+        }
+    }
+}
+
+impl From<TraceExporterError> for ExporterError {
+    // add code here
+    fn from(value: TraceExporterError) -> Self {
+        let code = match &value {
+            TraceExporterError::Builder(e) => match e {
+                BuilderErrorKind::InvalidUri => ExporterErrorCode::InvalidUrl,
+            },
+            TraceExporterError::Deserialization(_) => ExporterErrorCode::Serde,
+            TraceExporterError::Io(e) => match e.kind() {
+                IoErrorKind::InvalidData => ExporterErrorCode::InvalidData,
+                IoErrorKind::InvalidInput => ExporterErrorCode::InvalidInput,
+                IoErrorKind::ConnectionReset => ExporterErrorCode::ConnectionReset,
+                IoErrorKind::ConnectionAborted => ExporterErrorCode::ConnectionAborted,
+                IoErrorKind::ConnectionRefused => ExporterErrorCode::ConnectionRefused,
+                IoErrorKind::TimedOut => ExporterErrorCode::TimedOut,
+                IoErrorKind::AddrInUse => ExporterErrorCode::AddressInUse,
+                _ => ExporterErrorCode::IoError,
+            },
+            TraceExporterError::Network(e) => match e.kind() {
+                NetworkErrorKind::Body => ExporterErrorCode::HttpBodyFormat,
+                NetworkErrorKind::Canceled => ExporterErrorCode::ConnectionAborted,
+                NetworkErrorKind::ConnectionClosed => ExporterErrorCode::ConnectionReset,
+                NetworkErrorKind::MessageTooLarge => ExporterErrorCode::HttpBodyTooLong,
+                NetworkErrorKind::Parse => ExporterErrorCode::HttpParse,
+                NetworkErrorKind::TimedOut => ExporterErrorCode::TimedOut,
+                NetworkErrorKind::Unknown => ExporterErrorCode::NetworkUnknown,
+                NetworkErrorKind::WrongStatus => ExporterErrorCode::HttpWrongStatus,
+            },
+            TraceExporterError::Request(e) => {
+                let status: u16 = e.status().into();
+                if (400..500).contains(&status) {
```

Comment:
I know it _shouldn't_ happen, but Is it possible that we wind up with a `Request` error with a non-error status code? This if statement will return `exporterErrorCode::HttpServer` if the status code is < 400. If so, should we have a `ExporterErrorCode::Unknown` variant?

Reply:
Yep, by design it shouldn't but if there's some bug in the network library I think there's no harm in being prepared for that. I'll add another case in order to handle a generic error.